Userland/useradd.cpp

```diff
@@ -0,0 +1,145 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "home_dir", "home directory for the new user");
+    args_parser.add_arg("u", "uid", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "group", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "shell", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "name", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 0;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return 3;
+        }
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+    } else {
+        for (uid = BASE_UID; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return 1;
+        }
+    }
+
+    String shell = args.is_present("s") ? args.get("s") : String(DEFAULT_SHELL);
+    String home = args.is_present("d") ? args.get("d") : String("/");
+    String gecos = args.is_present("n") ? args.get("n") : String("Anonymous");
+
+    // Let's run a quick sanity check on username
+    if(strpbrk(username, "\\/123456789!@#$%^&*()~-_+=`:\n"))
+    {
+        perror("invalid character in username");
```

Comment:
`perror()` doesn't do what you think it does :smile:

Reply:
Oh shit that was meant to be `fprintf` damnit.